src/ML-SNAP/compute_sna_grid_local.cpp

```diff
@@ -0,0 +1,316 @@
+      for (int i = 0; i < ntypes; i++) {
+        int jelem = utils::inumeric(FLERR, arg[iarg + 2 + i], false, lmp);
+        if (jelem < 0 || jelem >= nelements)
+          error->all(FLERR, "Illegal compute sna/grid/local command");
+        map[i + 1] = jelem;
+      }
+      iarg += 2 + ntypes;
+    } else if (strcmp(arg[iarg], "bnormflag") == 0) {
+      if (iarg + 2 > narg) error->all(FLERR, "Illegal compute sna/grid/local command");
+      bnormflag = atoi(arg[iarg + 1]);
+      iarg += 2;
+    } else if (strcmp(arg[iarg], "wselfallflag") == 0) {
+      if (iarg + 2 > narg) error->all(FLERR, "Illegal compute sna/grid/local command");
+      wselfallflag = atoi(arg[iarg + 1]);
+      iarg += 2;
+    } else if (strcmp(arg[iarg], "switchinnerflag") == 0) {
+      if (iarg + 2 > narg) error->all(FLERR, "Illegal compute sna/grid/local command");
+      switchinnerflag = atoi(arg[iarg + 1]);
+      iarg += 2;
+    } else if (strcmp(arg[iarg], "sinner") == 0) {
+      iarg++;
+      if (iarg + ntypes > narg) error->all(FLERR, "Illegal compute sna/grid/local command");
+      memory->create(sinnerelem, ntypes + 1, "snap:sinnerelem");
+      for (int i = 0; i < ntypes; i++)
+        sinnerelem[i + 1] = utils::numeric(FLERR, arg[iarg + i], false, lmp);
+      sinnerflag = 1;
+      iarg += ntypes;
+    } else if (strcmp(arg[iarg], "dinner") == 0) {
+      iarg++;
+      if (iarg + ntypes > narg) error->all(FLERR, "Illegal compute sna/grid/local command");
+      memory->create(dinnerelem, ntypes + 1, "snap:dinnerelem");
+      for (int i = 0; i < ntypes; i++)
+        dinnerelem[i + 1] = utils::numeric(FLERR, arg[iarg + i], false, lmp);
+      dinnerflag = 1;
+      iarg += ntypes;
+    } else
+      error->all(FLERR, "Illegal compute sna/grid/local command");
+  }
+
+  if (switchinnerflag && !(sinnerflag && dinnerflag))
+    error->all(FLERR,
+               "Illegal compute sna/grid/local command: switchinnerflag = 1, missing sinner/dinner "
+               "keyword");
+
+  if (!switchinnerflag && (sinnerflag || dinnerflag))
+    error->all(FLERR,
+               "Illegal compute sna/grid/local command: switchinnerflag = 0, unexpected "
+               "sinner/dinner keyword");
+
+  snaptr = new SNA(lmp, rfac0, twojmax, rmin0, switchflag, bzeroflag, chemflag, bnormflag,
+                   wselfallflag, nelements, switchinnerflag);
+
+  ncoeff = snaptr->ncoeff;
+  nvalues = ncoeff;
+  if (quadraticflag) nvalues += (ncoeff * (ncoeff + 1)) / 2;
+  size_local_cols = size_local_cols_base + nvalues;
+}
+
+/* ---------------------------------------------------------------------- */
+
+ComputeSNAGridLocal::~ComputeSNAGridLocal()
+{
+  memory->destroy(radelem);
+  memory->destroy(wjelem);
+  memory->destroy(cutsq);
+  delete snaptr;
+
+  if (chemflag) memory->destroy(map);
+}
+
+/* ---------------------------------------------------------------------- */
+
+void ComputeSNAGridLocal::init()
+{
+  int count = 0;
```

Comment:
We try to minimize direct access to lists of items that are part of other classes (e.g. Modify here).
This check can be replaced with:
```
  if ((modify->get_compute_by_style("snal/grid/local").size() > 1) && (comm->me == 0))
    error->warning(FLERR, "More than one instance of compute orientorder/atom");
```
The `get...by_style()` functions return a vector with the pointers of classes and you can get its size/length directly.
Similar functions already exist for fixes, regions. More to come. You can also request `get...by_id()` and then you get just the one pointer (since ID are unique).

Reply:
I made this change. The only problem is, `sna/grid` matches `sna/grid/local`, not vice versa, so there is one false positive warning in the two examples. I assume this is related to matching style suffixes like `/kk`.

Reply:
You can request an exact match by using `^sna/grid$` instead of `sna/grid` and similarly `^sna/grid/local$`. If there will be future versions with suffixes, there can be more complex regular expressions (although this is a minimal matcher that does not support the entire standard regex syntax).

Reply:
That worked